### PR TITLE
feat(backend): implement flow scripts

### PR DESCRIPTION
**Pull request by uael**
This change introduces a script-like mechanism for flows to reduce the row size of the `queue` and `completed_job` tables, leading to improved performance.

## Implementation
A new `flow_version_lite` table, sharing the same `id` as `flow_version`, stores a modified flow value where `RawScript` modules are replaced with `FlowScript`. This optimized flow value is generated during the `FlowDependencies` job when the value is locked. A `FlowScript` is identified by a hash and a path: the hash is derived from the script lock file and content, while the path corresponds to the root flow path. When pushing a flow to the queue, the "lite" version is used if available; otherwise, the current behavior remains unchanged. A new `JobKind::FlowScript` is introduced to determine where to source the script.

## Backward compatibility
The new "lite" flow value version ins't used unless all workers are up to date. The env variable `DISABLE_FLOW_SCRIPT` can be used to disable the feature.

## Benchmarks
This optimization achieves up to a **45% speed improvement** for flows involving very large scripts.

## Open Questions
Flows deployed before this change will not automatically benefit from the optimization. While they can be manually re-deployed, a follow-up could involve applying the optimization in bulk to all existing flows.

----

> [!IMPORTANT]
> Introduces `FlowScript` to optimize flow handling by reducing data size in `queue` and `completed_job` tables, adding `flow_version_lite` table and updating backend logic.
> 
>   - **Behavior**:
>     - Introduces `flow_version_lite` table to store optimized flow versions with `FlowScript` replacing `RawScript`.
>     - Uses `flow_version_lite` when pushing flows to the queue if available.
>     - Adds `JobKind::FlowScript` to handle script sourcing.
>   - **Database**:
>     - Adds `flow_version_lite` and `flow_node` tables in `20241121133158_flow_node.up.sql`.
>     - Updates views in `20241121133158_flow_node.up.sql` and `20241121133158_flow_node.down.sql`.
>   - **Backend Logic**:
>     - Adds `FlowScript` variant to `FlowModuleValue` in `flows.rs`.
>     - Implements `create_flow_script()` in `worker_lockfiles.rs` for script creation.
>     - Updates `lock_modules()` in `worker_lockfiles.rs` to handle `FlowScript`.
>     - Modifies `handle_flow_dependency_job()` in `worker_lockfiles.rs` to update `flow_version_lite`.
>   - **Job Handling**:
>     - Adds `FlowScript` to `JobKind` and `JobPayload` in `jobs.rs`.
>     - Updates `push_next_flow_job()` in `worker_flow.rs` to handle `FlowScript`.
>     - Modifies `handle_queued_job()` in `worker.rs` to process `FlowScript` jobs.
>   - **SQL Queries**:
>     - Adds new queries for `flow_script` and `flow_version_lite` in `.sqlx` files.
>     - Updates existing queries to use `raw_flow_lock` instead of `raw_flow` in `worker_flow.rs` and `worker.rs`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=windmill-labs%2Fwindmill&utm_source=github&utm_medium=referral)<sup> for 82d651d722afdec9821c130b200dc724d91e6ae4. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: backend/windmill-queue/src/jobs.rs [hunk 1]
```diff
@@ -3299,6 +3294,31 @@ pub async fn push<'c, 'd, R: rsmq_async::RsmqConnection + Send + 'c>(
                 priority,
             )
         }
+        // TODO(uael): apply_preprocessor ?
```
Comment (uael) [inline backend/windmill-queue/src/jobs.rs]: TODO

File: backend/windmill-worker/src/worker_lockfiles.rs [hunk 2]
```diff
@@ -636,18 +636,26 @@ pub async fn handle_flow_dependency_job<R: rsmq_async::RsmqConnection + Send + S
             Error::InternalErr("Flow Dependency requires script hash (flow version)".to_owned())
         })?;
 
+        // TODO(uael): check min version here:
```
Comment (uael) [inline backend/windmill-worker/src/worker_lockfiles.rs]: TODO

Comment (uael) [reply backend/windmill-worker/src/worker_lockfiles.rs]: TODO: Generate two set of modules with the lock method, one locked but w/ flow script (as before), one locked but w flow scripts

File: backend/windmill-worker/src/worker_lockfiles.rs [hunk 3]
```diff
@@ -986,24 +986,97 @@ async fn lock_modules<'c>(
...
+    .execute(&mut *tx)
+    .await?;
+    Ok((tx, hash))
+}
+
+fn reduce<'c>(modules: &mut Vec<FlowModule>, flow_path: &str) {
+    use FlowModuleValue::*;
+    for module in &mut *modules {
+        let Ok(mut val) = serde_json::from_str::<FlowModuleValue>(module.value.get()) else { continue };
+        match &mut val {
```
[the remark this answers is not in the document]
Comment (uael) [reply backend/windmill-worker/src/worker_lockfiles.rs:1055]: Already handled

File: backend/windmill-worker/src/worker_lockfiles.rs [hunk 4]
```diff
@@ -986,24 +986,97 @@ async fn lock_modules<'c>(
...
+    workspace_id: &str,
+    lock: Option<&String>,
+    content: &String,
+) -> Result<(sqlx::Transaction<'c, sqlx::Postgres>, ScriptHash)> {
+    let hash = hash_script(lock, content);
+
+    sqlx::query!(
+        "INSERT INTO flow_script (workspace_id, hash, path, lock, content)
+        VALUES ($1, $2, $3, $4, $5)
+        ON CONFLICT DO NOTHING",
```
[the remark this answers is not in the document]
Comment (uael) [reply backend/windmill-worker/src/worker_lockfiles.rs]: Intended behavior

File: backend/windmill-worker/src/worker_lockfiles.rs [hunk 5]
```diff
@@ -986,24 +986,119 @@ async fn lock_modules<'c>(
...
+                    tag,
+                    language,
+                    custom_concurrency_key,
+                    concurrent_limit,
+                    concurrency_time_window_s,
+                    is_trigger,
+                };
+            },
+            ForloopFlow { modules, .. } | WhileloopFlow { modules, .. }  => {
+                tx = Box::pin(reduce(tx, &mut *modules, path, workspace_id)).await?;
```
[the remark this answers is not in the document]
Comment (uael) [reply backend/windmill-worker/src/worker_lockfiles.rs:1085]: Its a recursive async fn, hence pinning

**Review by rubenfiszel — APPROVED** [review]
Looks good, but improve silent errors

File: backend/windmill-common/src/flows.rs [hunk 6]
```diff
@@ -651,3 +692,82 @@ pub fn add_virtual_items_if_necessary(modules: &mut Vec<FlowModule>) {
...
+}
+
+/// Resolve modules recursively.
+pub async fn resolve_value(
+    e: &sqlx::PgPool,
+    workspace_id: &str,
+    value: &mut Box<JsonRawValue>,
+    with_code: bool,
+) {
+    let Ok(mut val) = serde_json::from_str::<FlowValue>(value.get()) else { return };
```
Comment (uael) [inline backend/windmill-common/src/flows.rs]: Forward error

File: backend/windmill-common/src/flows.rs [hunk 7]
```diff
@@ -651,3 +692,82 @@ pub fn add_virtual_items_if_necessary(modules: &mut Vec<FlowModule>) {
...
+/// Resolve module value recursively.
+pub async fn resolve_module(
+    e: &sqlx::PgPool,
+    workspace_id: &str,
+    value: &mut Box<JsonRawValue>,
+    with_code: bool,
+) {
+    use FlowModuleValue::*;
+
+    let Ok(mut val) = serde_json::from_str::<FlowModuleValue>(value.get()) else { return };
```
Comment (uael) [inline backend/windmill-common/src/flows.rs]: Forward error

File: backend/windmill-common/src/flows.rs [hunk 8]
```diff
@@ -651,3 +692,82 @@ pub fn add_virtual_items_if_necessary(modules: &mut Vec<FlowModule>) {
...
+            let (lock, content) = if !with_code {
+                (Some("...".to_string()), "...".to_string())
+            } else {
+                let Ok((lock, content)) = sqlx::query!(
+                    "SELECT lock, code AS \"code!: String\" FROM flow_node WHERE id = $1",
+                    id.0,
+                )
+                .fetch_one(e)
+                .await
+                .map(|record| (record.lock, record.code)) else { return };
```
Comment (uael) [inline backend/windmill-common/src/flows.rs]: Forward error

File: backend/windmill-worker/src/worker_lockfiles.rs [hunk 9]
```diff
@@ -977,24 +977,120 @@ async fn lock_modules<'c>(
...
+
+async fn reduce<'c>(
+    mut tx: sqlx::Transaction<'c, sqlx::Postgres>,
+    modules: &mut Vec<FlowModule>,
+    path: &str,
+    workspace_id: &str,
+) -> Result<sqlx::Transaction<'c, sqlx::Postgres>> {
+    use FlowModuleValue::*;
+    for module in &mut *modules {
+        let Ok(mut val) = serde_json::from_str::<FlowModuleValue>(module.value.get()) else { continue };
```
Comment (uael) [inline backend/windmill-worker/src/worker_lockfiles.rs]: Forward error

File: backend/windmill-common/src/flows.rs [hunk 10]
```diff
@@ -651,3 +692,82 @@ pub fn add_virtual_items_if_necessary(modules: &mut Vec<FlowModule>) {
...
+            for branch in branches {
+                for module in &mut branch.modules {
+                    Box::pin(resolve_module(e, workspace_id, &mut module.value, with_code)).await;
+                }
+            }
+        }
+        _ => {}
+    }
+    *value = to_raw_value(&val);
+}
```
Comment (uael) [inline backend/windmill-common/src/flows.rs]: New line eof

File: backend/migrations/20241121133158_flow_node.up.sql [hunk 11]
```diff
@@ -0,0 +1,23 @@
+-- Add up migration script here
+ALTER TYPE JOB_KIND ADD VALUE IF NOT EXISTS 'flowscript';
+
+-- Same as `flow_version` but with a "lite" value (raw scripts replaced by `flow_script`).
```
Comment (uael) [inline backend/migrations/20241121133158_flow_node.up.sql]: Nit: "`RawScript` modules replaced by `FlowScript` modules"

File: backend/migrations/20241121133158_flow_node.up.sql [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+CREATE TABLE flow_version_lite (
```
Comment (uael) [inline backend/migrations/20241121133158_flow_node.up.sql:5]: Grant all

File: backend/migrations/20241121133158_flow_node.up.sql [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    id BIGSERIAL PRIMARY KEY,
+    value JSONB,
+    FOREIGN KEY (id) REFERENCES flow_version (id) ON DELETE CASCADE
+);
+
+-- Either a script or a flow.
+CREATE TABLE flow_node (
```
Comment (uael) [inline backend/migrations/20241121133158_flow_node.up.sql:15]: Grant all

**Comment by uael** [comment]
> Looks good, but improve silent errors

Should be good now. Also added a feature flag `DISABLE_FLOW_SCRIPT`, and pinned activation version to `1.432.0`

**Pull request merged**